include/rcutils/error_handling.h

```diff
@@ -314,6 +314,93 @@ RCUTILS_PUBLIC
 void
 rcutils_reset_error(void);
 
+// Helper macro for nested macro expansion
+#define _RCUTILS_EXPAND(x) x
+
+/// Set the error message, as well as append the current file, line number, and previous error.
+/**
+ * If there is no previous error, has same behavior as RCUTILS_SET_ERROR_MSG.
+ * \param[in] msg The error message to be set.
+ */
+#define RCUTILS_SET_ERROR_MSG_AND_APPEND_PREV_ERROR(msg) \
+  do { \
+    if (rcutils_error_is_set()) { \
+      rcutils_error_string_t error_string = rcutils_get_error_string(); \
+      rcutils_reset_error(); \
+      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING( \
+        _RCUTILS_EXPAND(msg ":\n%s"), error_string.str); \
+    } else { \
+      RCUTILS_SET_ERROR_MSG(msg); \
+    } \
+  } while (0)
+
+/// Set the error message using a format string, format arguments, and previous error.
+/**
+ * This function sets the error message using the given format string, and appends and resets the
+ * latest error string.
+ * The resulting formatted string is silently truncated at RCUTILS_ERROR_MESSAGE_MAX_LENGTH.
+ *
+ * If there is no previous error, has same behavior as RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING.
+ *
+ * \param[in] format_string The string to be used as the format of the error message.
+ * \param[in] ... Arguments for the format string.
+ */
+#define RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING_AND_APPEND_PREV_ERROR(format_string, ...) \
+  do { \
+    if (rcutils_error_is_set()) { \
+      rcutils_error_string_t error_string = rcutils_get_error_string(); \
+      rcutils_reset_error(); \
+      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING( \
+        _RCUTILS_EXPAND(format_string ":\n%s"), __VA_ARGS__, error_string.str); \
+    } else { \
+      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(format_string, __VA_ARGS__); \
+    } \
+  } while (0)
+
+/// Write the given msg out to stderr, limiting the buffer size in the `fwrite`, appending the
+/// previous error.
+/**
+ * This will reset the previous error, if it exists.
+ * If there is no previous error, has same behavior as RCUTILS_SAFE_FWRITE_TO_STDERR.
+ */
+#define RCUTILS_SAFE_FWRITE_TO_STDERR_AND_APPEND_PREV_ERROR(msg) \
+  do { \
+    if (rcutils_error_is_set()) { \
+      rcutils_error_string_t error_string = rcutils_get_error_string(); \
+      rcutils_reset_error(); \
+      RCUTILS_SAFE_FWRITE_TO_STDERR_WITH_FORMAT_STRING( \
```

Comment:
Note that this macro has a hard length limit: https://github.com/ros2/rcutils/blob/5bd1230eb10595cec30168667adf03f48879b96c/include/rcutils/error_handling.h#L68

So as to avoid allocation. 

Using multiple calls to the non format version can be preferable. It allows for interleaving in multithreaded output but that's often a good trade off for writing and arbitrary amount of info to the screen without doing formatting. 

The format version is still appropriate when doing something other than concatenating strings. 

For this particular macro I think multiple calls might be better.